pkg/agent/storage/keyvalue_store_bbolt.go

```diff
@@ -0,0 +1,188 @@
+package storage
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"go.etcd.io/bbolt"
+)
+
+type bboltKeyValueStore struct {
+	logger     log.Logger
+	db         *bbolt.DB
+	bucketName string
+}
+
+func NewBBoltKeyValueStore(logger log.Logger, db *bbolt.DB, bucketName string) (*bboltKeyValueStore, error) {
+	err := db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists([]byte(bucketName))
+		if err != nil {
+			return fmt.Errorf("creating bucket: %w", err)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	m := &bboltKeyValueStore{
+		logger:     log.With(logger, "bucket", bucketName),
+		db:         db,
+		bucketName: bucketName,
+	}
+
+	return m, nil
+}
+
+func (s *bboltKeyValueStore) Get(key []byte) (value []byte, err error) {
+	if s == nil || s.db == nil {
+		return nil, errors.New("db is nil")
+	}
+
+	if err := s.db.View(func(tx *bbolt.Tx) error {
+		b := tx.Bucket([]byte(s.bucketName))
+		if b == nil {
+			return fmt.Errorf("%s bucket does not exist", s.bucketName)
+		}
+
+		value = b.Get(key)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return value, err
+}
+
+func (s *bboltKeyValueStore) Set(key, value []byte) error {
+	if s == nil || s.db == nil {
+		return errors.New("db is nil")
+	}
+
+	return s.db.Update(func(tx *bbolt.Tx) error {
+		b := tx.Bucket([]byte(s.bucketName))
+		if b == nil {
+			return fmt.Errorf("%s bucket does not exist", s.bucketName)
+		}
+
+		if value != nil {
+			if err := b.Put(key, value); err != nil {
+				return err
```

Comment:
Potentially worth wrapping this err with the key included? 🤷‍♀️  Same q in `Delete` below

Reply:
Agree!